File: module-07-transport-layer/udp_echo_server.js
/**
 * Module 07 — JS/Node port of udp_echo_server.py. Same deliberate-drop
 * demonstration using Node's `dgram` module (UDP).
 * Binds to 127.0.0.1 only. See ../SAFETY.md.
 */
const dgram = require("dgram");

const HOST = "127.0.0.1";
const PORT = 9019; // different port from the Python variant

const DROP_RATE = 0.3;

// A small seeded PRNG so this demo's drop pattern is reproducible run to
// run, same spirit as Python's random.seed(7) -- Math.random() can't be
// seeded directly in Node, so a tiny linear congruential generator stands in.
function makeSeededRandom(seed) {
  let state = seed;
  return () => {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    return state / 0x7fffffff;
  };
}
const rng = makeSeededRandom(7);

const server = dgram.createSocket("udp4");
let count = 0;

server.on("message", (msg, rinfo) => {
  count++;
  if (rng() < DROP_RATE) {
    console.log(`  received #${count}: ${JSON.stringify(msg.toString())} -- DROPPING reply on purpose (simulated loss)`);
    return;
  }
  console.log(`  received #${count}: ${JSON.stringify(msg.toString())} -- echoing back`);
  server.send(msg, rinfo.port, rinfo.address);
});

server.bind(PORT, HOST, () => {
  console.log(`UDP echo server listening on ${HOST}:${PORT}`);
  console.log(`Deliberately dropping ~${Math.round(DROP_RATE * 100)}% of replies -- watch the client's gaps.`);
});

File: module-07-transport-layer/udp_echo_server.py
"""
Module 07 -- a UDP echo server that DELIBERATELY drops some replies, on
purpose, to make UDP's lack of guarantees visible instead of theoretical.

UDP itself doesn't drop packets under normal conditions -- your OS and
network hardware are usually reliable. What UDP lacks is any BUILT-IN
mechanism to notice or fix loss if it happens (congestion, a flaky link, a
saturated queue). This script fakes that "if it happens" condition on
purpose, at a fixed rate, so you can observe what UDP does about it:
nothing at all. No retransmission, no gap detection, no ordering
guarantee -- that's entirely on whatever's built on top of UDP to handle
(DNS does a version of this with retry+timeout; Module 08 builds a real one).

Binds to 127.0.0.1 only. See ../SAFETY.md.
"""
import random
import socket

HOST, PORT = "127.0.0.1", 9009
DROP_RATE = 0.3  # drop 30% of replies on purpose


def main():
    random.seed(7)  # reproducible drop pattern for this demo
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as server:
        server.bind((HOST, PORT))
        print(f"UDP echo server listening on {HOST}:{PORT}")
        print(f"Deliberately dropping ~{int(DROP_RATE * 100)}% of replies -- watch the client's gaps.")

        count = 0
        while True:
            data, addr = server.recvfrom(1024)
            count += 1
            if random.random() < DROP_RATE:
                print(f"  received #{count}: {data!r} -- DROPPING reply on purpose (simulated loss)")
                continue
            print(f"  received #{count}: {data!r} -- echoing back")
            server.sendto(data, addr)


if __name__ == "__main__":
    main()
